File: splay.cpp
#include <cstdio>
#include <cstring>
using namespace std;
#define infinity 1000000000
#define limitsize 1000
struct TreeNode
{
	int left, right, father;
	int count;
	int key;
};
struct Tsplay
{
	int root, top;
	TreeNode tree[limitsize];
	
	void initialize()
	{
		root = top = 0;
		memset(tree, 0, sizeof(tree));
	}
	
	void left_rotate(int x)
	{
		int y = tree[x].father, z = tree[x].left;
		if (y == tree[tree[y].father].left)
			tree[tree[y].father].left = x;
		else
			tree[tree[y].father].right = x;
		tree[x].father = tree[y].father;
		tree[x].left = y;
		tree[y].right = z;
		tree[y].father = x;
		tree[z].father = y;
	}
	
	void right_rotate(int x)
	{
		int y = tree[x].father, z = tree[x].right;
		if (y == tree[tree[y].father].left)
			tree[tree[y].father].left = x;
		else
			tree[tree[y].father].right = x;
		tree[x].father = tree[y].father;
		tree[x].right = y;
		tree[y].left = z;
		tree[y].father = x;
		tree[z].father = y;
	}
	
	void splay(int x)
	{
		while (tree[x].father)
		{
			int y = tree[x].father, z = tree[y].father;
			if (z == 0)
			{
				if (x == tree[y].left)
					right_rotate(x);
				else
					left_rotate(x);
			}
			else
			{
				if (x == tree[y].left)
				{
					if (y == tree[z].left)
					{
						right_rotate(y);
						right_rotate(x);
					}
					else
					{
						right_rotate(x);
						left_rotate(x);
					}
				}
				else
				{
					if (y == tree[z].left)
					{
						left_rotate(x);
						right_rotate(x);
					}
					else
					{
						left_rotate(y);
						left_rotate(x);
					}
				}
			}
			
		}
		root = x;
	}
	
	int find(int x, int key)
	{
		while (tree[x].key != key)
			if (key < tree[x].key)
			{
				if (tree[x].left == 0) break;
				x = tree[x].left;
			}
			else
			{
				if (tree[x].right == 0) break;
				x = tree[x].right;
			}
		return x;
	}

	int maximum(int x)
	{
		splay(find(x, infinity));
		return root;
	}
	
	int minimum(int x)
	{
		splay(find(x, -infinity));
		return root;
	}
	
	int maximum_key(int x)
	{
		return tree[maximum(x)].key;
	}
	
	int minimum_key(int x)
	{
		return tree[minimum(x)].key;
	}
	
	int join(int x1, int x2)
	{
		if (x1 == 0) return x2;
		if (x2 == 0) return x1;
		int x = maximum(x1);
		tree[x].right = x2; tree[x2].father = x;
		return root;
	}	

	void split(int x, int& x1, int& x2)
	{
		splay(x);
		tree[x1 = tree[x].left].father = 0; tree[root].left = 0;
		tree[x2 = tree[x].right].father = 0; tree[root].right = 0;
	}
	
	void insert(int key)
	{
		tree[++ top].key = key; tree[top].count = 1;
		tree[top].left = tree[top].right = tree[top].father = 0;
		if (root == 0)
		{
			tree[root = top].father = 0;
			return;
		}
		splay(find(root, key));
		if (tree[root].key == key)
		{
			++ tree[root].count;
			return;
		}
		if (key < tree[root].key)
		{
			tree[top].left = tree[root].left;
			if (tree[top].left != 0)
				tree[tree[top].left].father = top;
			tree[root].left = 0;
			tree[top].right = root;
		}
		else
		{
			tree[top].right = tree[root].right;
			if (tree[top].right != 0)
				tree[tree[top].right].father = top;
			tree[root].right = 0;
			tree[top].left = root;
		}
		tree[root].father = top;
		root = top;
	}
	
	void erase(int key)
	{
		int x = find(root, key);
		if (key != tree[x].key) return;
		if (tree[x].count > 1)
		{
			-- tree[x].count;
			return;
		}
		int s1, s2;
		split(x, s1, s2);
		root = join(s1, s2);
	}
	int prev(int key)
	{
		int p = find(root, key);
		splay(p);
		if (tree[p].left == 0) return tree[p].key;
		p = tree[p].left;
		return tree[maximum(p)].key;
	}
	
	int succ(int key)
	{
		int p = find(root, key);
		splay(p);
		if (tree[p].right == 0) return tree[p].key;
		p = tree[p].right;
		return tree[minimum(p)].key;
	}
	
	bool exist(int key)
	{
		return (tree[find(root, key)].key == key);
	}
};

int main()
{
	return 0;
}

File: PascalLibrary/Collection/splay.pas
program splay;
const
	maxn = 20000;
	infinity = 1000000000;
type
	treenode = record
		left,right,father,count:longint;
		key:longint;
	end;
	tsplay = object
		root,top:longint;
		tree:array[1..maxn] of treenode;
		
		procedure initialize;
		procedure left_rotate(x:longint);
		procedure right_rotate(x:longint);
		function find(t:longint; key:longint):longint;
		procedure splay(x:longint);
		
		procedure insert(key:longint);
		procedure delete(key:longint);
		function join(x1,x2:longint):longint;
		procedure split(x:longint; var s1,s2:longint);
		
		function minimum(x:longint):longint;
		function getminimum(x:longint):longint;
		function maximum(x:longint):longint;
		function getmaximum(x:longint):longint;
		function exist(key:longint):boolean;
		function prev(key:longint):longint;
		function succ(key:longint):longint;
	end;

procedure tsplay.initialize;
 begin
 	root:=0; top:=0;
 	fillchar(tree,sizeof(tree),0);
 end;

procedure tsplay.left_rotate(x:longint);
 var
 	y,z:longint;
 begin
 	y:=tree[x].father; z:=tree[x].left;
 	if (y = tree[tree[y].father].left) then tree[tree[y].father].left:=x
 		else tree[tree[y].father].right:=x;
 	
 	tree[x].father:=tree[y].father;
 	tree[x].left:=y;
 	tree[y].right:=z;
 	tree[y].father:=x;
 	tree[z].father:=y;
 end;
 
procedure tsplay.right_rotate(x:longint);
 var
 	y,z:longint;
 begin
 	y:=tree[x].father; z:=tree[x].right;
 	if (y = tree[tree[y].father].left) then tree[tree[y].father].left:=x
 		else tree[tree[y].father].right:=x;
 	
 	tree[x].father:=tree[y].father;
 	tree[x].right:=y;
 	tree[y].left:=z;
 	tree[y].father:=x;
 	tree[z].father:=y;
 end;

procedure tsplay.splay(x:longint);
 var
 	y,z:longint;
 begin
 	while (tree[x].father > 0) do begin
 		y:=tree[x].father;
 		z:=tree[y].father;
 		if (z = 0) then begin
 			if (x = tree[y].left) then right_rotate(x) else left_rotate(x);
 		end else begin
 			if (x = tree[y].left) then begin
 				if (y = tree[z].left) then begin
 					right_rotate(y);
 					right_rotate(x);
 				end else begin
 					right_rotate(x);
 					left_rotate(x);
 				end;
 			end else begin
 				if (y = tree[z].left) then begin
 					left_rotate(x);
 					right_rotate(x);
 				end else begin
 					left_rotate(y);
 					left_rotate(x);
 				end;
 			end;
 		end;
 	end;
 	root:=x;
 end;

function tsplay.find(t:longint; key:longint):longint;
 begin
 	while (tree[t].key <> key) do
 		if (key < tree[t].key) then begin
 			if (tree[t].left = 0) then break;
 			t:=tree[t].left;
 		end else begin
 			if (tree[t].right = 0) then break;
 			t:=tree[t].right;
 		end;
 	find:=t;
 end;

procedure tsplay.insert(key:longint);
 begin
 	inc(top);
 	tree[top].key:=key; tree[top].count:=1;
 	tree[top].left:=0; tree[top].right:=0; tree[top].father:=0;
 	if (root = 0) then begin
 		root:=top;
 		exit;
 	end;
 	splay(find(root,key));
 	if (tree[root].key = key) then begin
 		inc(tree[root].count);
 		exit;
 	end;
 	if (key < tree[root].key) then begin
 		tree[top].left:=tree[root].left;
 		if (tree[top].left <> 0) then tree[tree[top].left].father:=top;
 		tree[root].left:=0;
 		tree[top].right:=root;
 	end else begin
 		tree[top].right:=tree[root].right;
 		if (tree[top].right <> 0) then tree[tree[top].right].father:=top;
 		tree[root].right:=0;
 		tree[top].left:=root;
 	end;
 	tree[root].father:=top;
 	root:=top;
 end;

procedure tsplay.delete(key:longint);
 var
 	t,s1,s2:longint;
 begin
 	t:=find(root,key);
 	if (key <> tree[t].key) then exit;
 	if (tree[t].count > 1) then begin
 		dec(tree[t].count);
 		exit;
 	end;
 	split(t,s1,s2);
 	root:=join(s1,s2);
 end;

function tsplay.join(x1,x2:longint):longint;
 var
 	p:longint;
 begin
	if (x1 = 0) then exit(x2);
	if (x2 = 0) then exit(x1); 
	p:=maximum(x1); tree[p].right:=x2; tree[x2].father:=p;
	join:=root;
 end;

procedure tsplay.split(x:longint; var s1,s2:longint);
 begin
 	splay(x);
 	s1:=tree[x].left; tree[s1].father:=0; tree[root].left:=0;
 	s2:=tree[x].right; tree[s2].father:=0; tree[root].right:=0;
 end;

function tsplay.minimum(x:longint):longint;
 begin
 	splay(find(x,-infinity));
 	minimum:=root;
 end;

function tsplay.maximum(x:longint):longint;
 begin
 	splay(find(x,infinity));
 	maximum:=root;
 end;

function tsplay.getmaximum(x:longint):longint;
 begin
 	getmaximum:=tree[maximum(x)].key;
 end;

function tsplay.getminimum(x:longint):longint;
 begin
 	getminimum:=tree[minimum(x)].key;
 end;

function tsplay.prev(key:longint):longint;
 var
	p:longint;
 begin
	p:=find(root,key);
	splay(p);
	if (tree[p].left = 0) then exit(tree[p].key);
	p:=tree[p].left;
	prev:=tree[maximum(p)].key;
 end;

function tsplay.succ(key:longint):longint;
 var
	p:longint;
 begin
	p:=find(root,key);
	splay(p);
	if (tree[p].right = 0) then exit(tree[p].key);
	p:=tree[p].right;
	succ:=tree[minimum(p)].key;
 end;

function tsplay.exist(key:longint):boolean;
 begin
 	exist:=tree[find(root,key)].key = key;
 end;

var
	splaytree:tsplay;

begin
	splaytree.initialize;
end.

{
procedure tsplay.delete(key:longint);
 var
 	t:longint;
 begin
 	t:=find(root,key);
 	if (key <> tree[t].key) then exit;
 	if (tree[t].count > 1) then begin
 		dec(tree[t].count);
 		exit;
 	end;
 	splay(t);
 	if (tree[t].left = 0) then begin
 		root:=tree[t].right;
 		tree[root].father:=0;
 	end else begin
 		tree[tree[t].left].father:=0;
 		root:=find(tree[t].left,infinity);
 		splay(root);
 		tree[root].right:=tree[t].right;
 		tree[tree[root].right].father:=root;
 	end;
 end;
}
 
{
	assign(input,'bst.in'); reset(input);
	assign(output,'splay.out'); rewrite(output);
	main;
	close(input);
	close(output);
}


{
	splaytree.insert(1);
	splaytree.insert(3);
	splaytree.insert(2);
	splaytree.insert(10);
	splaytree.insert(0);
	writeln(splaytree.getmaximum(splaytree.root));
	writeln(splaytree.getminimum(splaytree.root));
	
	splaytree.delete(10);
	writeln(splaytree.getmaximum(splaytree.root));
	writeln(splaytree.getminimum(splaytree.root));
	
	splaytree.delete(0);
	writeln(splaytree.getmaximum(splaytree.root));
	writeln(splaytree.getminimum(splaytree.root));
}

{
procedure main;
 var
 	n,i,k,key:longint;
 begin
 	readln(n);
 	for i:=1 to n do begin
 		read(k);
 		if (k <= 3) or (k = 6) or (k = 7) then begin
 			readln(key);
 			if (k = 1) then splaytree.insert(key);
 			if (k = 2) then splaytree.delete(key);
 			if (k = 3) then writeln(splaytree.exist(key));
			if (k = 6) then writeln(splaytree.prev(key));
			if (k = 7) then writeln(splaytree.succ(key));
 		end else begin
 			if (k = 8) then writeln(splaytree.getmaximum(splaytree.root));
 			if (k = 9) then writeln(splaytree.getminimum(splaytree.root));
 			readln;
 		end;
 	end;
 end;
}